Add docs for Payment Handler API (#23270)
mdn/content: files/en-us/web/api/payment_handler_api/index.md

slug: Web/API/Payment_Handler_API
page-type: web-api-overview
tags:
  - API
  - Experimental
  - Landing
  - Overview
  - Payment Handler API
  - Reference
browser-compat: api.PaymentRequestEvent
---

{{DefaultAPISidebar("Payment Handler API")}}{{securecontext_header}}{{SeeCompatTable}}

The Payment Handler API provides a standardized set of functionality for web applications to directly handle payments, rather than having to be redirected to a separate site for payment handling.

When a merchant website initiates payment via the {{domxref("Payment Request API", "Payment Request API", "", "nocode")}}, the Payment Handler API handles discovery of applicable payment apps, presenting them as choices to the user, opening a payment handler window once a choice has been made to allow the user to enter their payment details, and handling the payment transaction with the payment app.

Communication with payment apps (authorization, passing of payment credentials) is handled via Service Workers.

## Concepts and usage

On a merchant website, a payment request is initiated by the construction of a new {{domxref("PaymentRequest")}} object:

```js
const request = new PaymentRequest([{

>   supportedMethods: 'https://bobpay.xyz/pay'

Author: Yeah, makes sense. I've updated all instances.

}], {
  total: {
    label: 'total',
    amount: { value: '10', currency: 'USD' }
  }
});
```

> The `supportedMethods` property specifies one or more URLs representing the payment methods supported by the merchant.

Author: Gotcha, I've updated the text, and added this in as an additional example.

### Making payment apps available

> In supporting browsers, the Payment Request API starts off the process of installing payment apps that can use the payment methods by requesting a [payment method manifest](TBD) file from each URL. Those manifests are returned either directly, or indirectly via a {{httpheader("Link")}} header if stored somewhere else (see [Optionally route the browser to find the payment method manifest in another location](https://web.dev/setting-up-a-payment-method/#optionally-route-the-browser-to-find-the-payment-method-manifest-in-another-location) for details).

Author: Ah ha, thanks for the clarification here. I was getting a bit confused by when exactly the payment app is installed. In my next commit, I've made the change suggested above, but I've also made some significant changes to the rest of the explanation, to explain this whole part of the process a bit more clearly. Please can you have anoither read of this section and let me know if it works?

> A payment method manifest should be called `payment-manifest.json`, and be structured like this:

Author: Cool, this text is useful, so I've added it in (with a little tweaking).

I've also made it clear that you are not tied to that exact name for the manifest.

```json
{
  "default_applications": ["https://bobpay.xyz/manifest.json"],
  "supported_origins": [
    "https://alicepay.friendsofalice.example"
  ]
}
```

- `default_applications` tells the browser where to find the default payment app that can use the BobPay payment method if it doesn't already have one installed.

> - `supported_origins` tells the browser where to find other payment apps that it is permitted to use to handle the BobPay payment if needed.

Author: I made those changes, but I also thought it would be good to add another sentence clarifying how the browser uses the supported origins. I ended up with this:

> - `supported_origins` tells the browser what other payment apps are permitted to handle the BobBucks payment if needed. If they are already installed on the device, they will be presented to the user as alternative payment options alongside the default application.

Does that work?

> - `supported_origins` tells the browser where to find other payment apps that it is permitted to use to handle the BobPay payment if needed.

Reviewer: Yes, that's perfect! 👍

> From that manifest, the browser gets the URL of the [web app manifest](/en-US/docs/Web/Manifest) of the supported payment app, which should be called `manifest.json` and look something like this:

Author: I've updated it so that it is clear that they can be called whatever you like.

When the payment has been handled, {{domxref("PaymentRequestEvent.respondWith()")}} is used to pass the payment result back to the merchant website.

See [Receive a payment request event from the merchant](https://web.dev/orchestrating-payment-transactions/#receive-payment-request-event) for more details of this stage.

### Managing payment app functionality

Once a payment app service worker is registered, you can use the service worker's {{domxref("PaymentManager")}} instance (accessed via {{domxref("ServiceWorkerRegistration.paymentManager")}}) to manage various aspects of the payment app's functionality.

For example:

```js
navigator.serviceWorker.register("serviceworker.js")
  .then(registration => {
    registration.paymentManager.userHint = "Card number should be 16 digits";

    registration.paymentManager.enableDelegations(['shippingAddress', 'payerName']) 
      .then(() => {
          // ...
      });

    // ...

  });
```

> - {{domxref("PaymentManager.userHint")}} is used to provide a hint for the browser to display along with the payment app's name and icon in the Payment Handler UI.

Author: Noted. We'll keep an eye on this; I'm happy to update the docs when required.

> - {{domxref("PaymentManager.enableDelegations()")}} is used to delegate responsibility for providing various parts of the required payment information to the payment app rather than collecting it from within the merchant website.

Author: I've made the change to the paragraph, similar to suggested, but I think on the `enableDelegations()` page it might be good to walk through an actual example, like you have done above. However, I need a bit more info here — where do you pass `enableShipping: true` into the API? I couldn't find mention of that on existing MDN docs, or in the Payment Request/Handler specs.

> - {{domxref("PaymentManager.enableDelegations()")}} is used to delegate responsibility for providing various parts of the required payment information to the payment app rather than collecting it from within the merchant website.

Reviewer: I'm so sorry about the confusion! That was my fault. I should've written **`requestShipping: true`** instead.

The `requestShipping: true` is passed into the 3rd parameter of the `new PaymentRequest()` constructor, like so:

```js
new PaymentRequest(methods, details, {requestShipping: true});
```

The `requestShipping: true` option will be placed back into the spec in [the pull request to re-enable shipping address](https://github.com/w3c/payment-request/pull/996/files#diff-0eb547304658805aad788d320f10bf1f292797b5e6d745a3bf617584da017051R465).

> - {{domxref("PaymentManager.enableDelegations()")}} is used to delegate responsibility for providing various parts of the required payment information to the payment app rather than collecting it from within the merchant website.

Author: No worries! We still have `requestShipping` documented on MDN, so this makes sense. I've added an example explanation to the `enableDelegations()` page, and I think it reads pretty well.

From the payment method manifest, the browser gets the URL of the default payment apps' [web app manifest](/en-US/docs/Web/Manifest) files, which can be called whatever you like, and look something like this:

```json
{
  "name": "Pay with BobBucks",
  "short_name": "BobBucks",
  "description": "This is an example of the Payment Handler API.",
  "icons": [
    {
      "src": "images/manifest/icon-192x192.png",
      "sizes": "192x192",
      "type": "image/png"
    },
    {
      "src": "images/manifest/icon-512x512.png",
      "sizes": "512x512",
      "type": "image/png"
    }
  ],
  "serviceworker": {
    "src": "service-worker.js",
    "scope": "/",
    "use_cache": false
  },
  "start_url": "/",
  "display": "standalone",
  "theme_color": "#3f51b5",
  "background_color": "#3f51b5",
  "related_applications": [
    {
      "platform": "play",
      "id": "com.example.android.samplepay",
      "min_version": "1",
      "fingerprints": [
        {
          "type": "sha256_cert",
          "value": "4C:FC:14:C6:97:DE:66:4E:66:97:50:C0:24:CE:5F:27:00:92:EE:F3:7F:18:B3:DA:77:66:84:CD:9D:E9:D2:CB"
        }
      ]
    }
  ]
}
```

> When the {{domxref("PaymentRequest.show()")}} method is invoked by the merchant app in response to a user gesture, the browser uses the information found in each manifest to present the payment apps to the user in the browser-provided Payment Request UI ([`name`](/en-US/docs/Web/Manifest/name), [`icons`](/en-US/docs/Web/Manifest/icons)).

Reviewer 2: ```suggestion
When the {{domxref("PaymentRequest.show()")}} method is invoked by the merchant app in response to a user gesture, the browser uses the [`name`](/en-US/docs/Web/Manifest/name) and [`icons`](/en-US/docs/Web/Manifest/icons information found in each manifest to present the payment apps to the user in the browser-provided Payment Request UI ).
```

```js
async function checkCanMakePayment() {

  // ...

  const canMakePayment = await request.canMakePayment();
  if (!canMakePayment) {
    // Fallback to other means of payment or hide the button.
  }
}
```

The Payment Handler API adds an additional mechanism to prepare for handling a payment. The {{domxref("ServiceWorkerGlobalScope.canmakepayment_event", "canmakepayment")}} event is fired on a payment app's service worker to check whether it is ready to handle a payment. Specifically, it is fired when the merchant website calls {{domxref("PaymentRequest.PaymentRequest", "new PaymentRequest()")}}. The service worker can then use the {{domxref("CanMakePaymentEvent.respondWith()")}} method to respond appropriately:

```js
self.addEventListener("canmakepayment", e => {
  e.respondWith(new Promise((resolve, reject) => {
    someAppSpecificLogic()
    .then(result => { resolve(result); })
    .catch(error => { reject(error); });
  }));
});
```

> The promise returned by `respondWith()` resolves with a boolean to signal that it is ready to handle a payment request (`true`), or not (`false`).

Reviewer 2: ```suggestion
The promise returned by `respondWith()` resolves with a boolean to signal that it is ready to handle a payment request: (`true`), or not (`false`).
```

- {{domxref("PaymentManager.enableDelegations()")}} is used to delegate responsibility for providing various parts of the required payment information to the payment app rather than collecting it from the browser (for example, via autofill).

## Interfaces

- {{domxref("CanMakePaymentEvent")}}
  - : The event object for the {{domxref("ServiceWorkerGlobalScope.canmakepayment_event", "canmakepayment")}} event, fired on a payment app's service worker when it has been successfully registered to signal that it is ready to handle payments.
- {{domxref("PaymentManager")}}
  - : Used to manage various aspects of payment app functionality. Accessed via the {{domxref("ServiceWorkerRegistration.paymentManager")}} property.
- {{domxref("PaymentRequestEvent")}}
  - : The event object for the {{domxref("ServiceWorkerGlobalScope.paymentrequest_event", "paymentrequest")}} event, fired on a payment app's service worker when a payment flow has been initiated on the merchant website via the {{domxref("PaymentRequest.show()")}} method.

## Extensions to other interfaces

- {{domxref("ServiceWorkerGlobalScope.canmakepayment_event", "canmakepayment")}} event

>   - : Fired on a payment app's {{domxref("ServiceWorkerGlobalScope")}} when it has been successfully registered to signal that it is ready to handle payments.

Reviewer 2: ```suggestion
  - : Fired on a payment app's {{domxref("ServiceWorkerGlobalScope")}} when it has been successfully registered, to signal that it is ready to handle payments.
```

From the payment method manifest, the browser gets the URL of the default payment apps' [web app manifest](/en-US/docs/Web/Manifest) files, which can be called whatever you like, and look something like this:

```json
{
  "name": "Pay with BobBucks",
  "short_name": "BobBucks",
  "description": "This is an example of the Payment Handler API.",
  "icons": [
    {
      "src": "images/manifest/icon-192x192.png",
      "sizes": "192x192",
      "type": "image/png"
    },
    {
      "src": "images/manifest/icon-512x512.png",
      "sizes": "512x512",
      "type": "image/png"
    }
  ],
  "serviceworker": {
    "src": "service-worker.js",
    "scope": "/",
    "use_cache": false
  },
  "start_url": "/",
  "display": "standalone",
  "theme_color": "#3f51b5",
  "background_color": "#3f51b5",
  "related_applications": [
    {
      "platform": "play",
      "id": "com.example.android.samplepay",
      "min_version": "1",
      "fingerprints": [
        {
          "type": "sha256_cert",
          "value": "4C:FC:14:C6:97:DE:66:4E:66:97:50:C0:24:CE:5F:27:00:92:EE:F3:7F:18:B3:DA:77:66:84:CD:9D:E9:D2:CB"
        }
      ]
    }
  ]
}
```

> It first uses the information found in each manifest to present the payment apps to the user in the browser-provided Payment Request UI in the merchant app ([`name`](/en-US/docs/Web/Manifest/name), [`icons`](/en-US/docs/Web/Manifest/icons)).
> 
> Next, the {{domxref("PaymentRequest.show()")}} method should be invoked by the merchant app in response to a user gesture:
> 
> - If there are multiple payment app options, a list of options is presented to the user from them to choose from.
> - If there is only one payment app option, the user gesture is still required to continue — along the lines of clicking a "proceed to payment" button.
> 
> This starts the payment flow, which causes the browser to Just-In-Time (JIT) install the web app, registering the service worker specified in [`serviceworker`](/en-US/docs/Web/Manifest/serviceworker) so it can handle the payment.

Reviewer: This text can be confusing when trying to understand the timing of when the browser shows the UI or launches an app, because it first talks about browser-provided UI and next goes into details of `PaymentRequest.show()`.  However, it is `PaymentRequest.show()` that triggers showing any UI. The browser does not show any UI before `PaymentRequest.show()` is called.

It may be easier to understand if `PaymentRequest.show()` is mentioned first.

Here is a possible rephrasing. Please feel free to change it any way that you feel fit.

```suggestion
When the {{domxref("PaymentRequest.show()")}} method is invoked by the merchant app in response to a user gesture, the browser uses the information found in each manifest to present the payment apps to the user in the browser-provided Payment Request UI ([`name`](/en-US/docs/Web/Manifest/name), [`icons`](/en-US/docs/Web/Manifest/icons)).

If there are multiple payment app options, a list of options is presented to the user for them to choose one. Selecting one payment app will start the payment flow, which causes the browser to Just-In-Time (JIT) install the web app if necessary, registering the service worker specified in [`serviceworker`](/en-US/docs/Web/Manifest/serviceworker) so it can handle the payment.

If there is only one payment app option, the {{domxref("PaymentRequest.show()")}} method will launch this payment app, JIT-installing it if necessary. (This is an optimization to avoid presenting the user with a list that contains only one payment app choice.)
```

Author: Makes sense, thanks. I think we've nailed this now. I mostly used your text with a few tweaks.

> **Note:** If [`prefer_related_applications`](/en-US/docs/Web/Manifest/prefer_related_applications) is set to `true` in the payment app manifest, the browser will launch the platform-specific payment app specified in [`related_applications`](/en-US/docs/Web/Manifest/related_applications) to handle the payment (if it is available) instead of the web payment app.

See [Serve a web app manifest](https://web.dev/setting-up-a-payment-method/#step-3-serve-a-web-app-manifest) for more details.

### Checking whether the payment app is ready to pay with

The Payment Request API's {{domxref("PaymentRequest.canMakePayment()")}} method returns `true` if a payment app is available on the customer's device, meaning that a payment app that supports the payment method is discovered, and that the platform-specific payment app is installed, or the web-based payment app is ready to be registered.

```js
async function checkCanMakePayment() {

  // ...

  const canMakePayment = await request.canMakePayment();
  if (!canMakePayment) {
    // Fallback to other means of payment or hide the button.
  }
}
```

> The Payment Handler API adds an additional mechanism to prepare for handling a payment. The {{domxref("ServiceWorkerGlobalScope.canmakepayment_event", "canmakepayment")}} event is fired on a payment app's service worker to check whether it is ready to handle a payment. Specifically, it is fired when the merchant website calls {{domxref("PaymentRequest.PaymentRequest", "new PaymentRequest()")}}. The service worker can then use the {{domxref("CanMakePaymentEvent.respondWith()")}} method to respond appropriately:

Reviewer: Perfection! 👩‍🍳

Author: w00t!
